t/mojo-oauth2.t

```diff
@@ -47,129 +54,151 @@ $t->ua->max_redirects(1);
-$t->get_ok('/api/user/profile' =>
-    {Authorization => 'Bearer ' . $access_data->{access_token}})->status_is(200)
-  ->json_is('/login' => $oauth_login);
-
-# Should be able to use the refresh token to get a new access token
-$t->post_ok(
-  '/oauth/access_token' => {Referer => $referer} => form => {
-    client_id     => $oauth_client->{client_id},
-    client_secret => $oauth_client->{secret},
-    refresh_token => $access_data->{refresh_token},
-    grant_type    => 'refresh_token',
-    redirect_uri  => '/oauth/redirect',
-  }
-)->status_is(200)->json_has('/access_token', 'Has access token')
-  ->json_has('/refresh_token', 'Has refresh token')
-  ->json_has('/token_type',    'Has token type');
-
-$access_data = $t->tx->res->json;
+# Make sure we can do a normal login and access our profile data
+my $access_data = login(
+  t        => $t,
+  client   => $oauth_client,
+  user     => $oauth_user,
+  login    => $oauth_login,
+  password => $oauth_password
+);
 
 $t->get_ok('/api/user/profile' =>
     {Authorization => 'Bearer ' . $access_data->{access_token}})->status_is(200)
-  ->json_is('/login' => $oauth_login);
+  ->json_is('/login' => $oauth_user->email);
+
+# Trying to access using an OAuth2 user without 'api:modify' scope on their
+# account should also fail.
+$t->get_ok("/rest/user/"
+    . $oauth_user->id =>
+    {Authorization => 'Bearer ' . $access_data->{access_token}})->status_is(401);
+
+# Login to the API enabled OAuth2 client
+$access_data = login(
+  t        => $t,
+  client   => $oauth_api_client,
+  user     => $oauth_user,
+  login    => $oauth_login,
+  password => $oauth_password
+
+);
+
+# We should also be able to call any legacy REST API method
+# using the OAuth2 access token in the header.
+# /rest/user/<id> is a good example as it requires login
+$t->get_ok("/rest/user/"
+    . $oauth_user->id =>
+    {Authorization => 'Bearer ' . $access_data->{access_token}})->status_is(200)
+  ->json_is('/users/0/email' => $oauth_user->email);
+
+# Try to use /rest/user/<id> without logging in using OAuth2 token
+# should result in an access error.
+$t->get_ok("/rest/user/" . $oauth_user->id)->status_is(401);
 
-# API should fail if user is disabled
+# User profile API call should fail if user is disabled
 $oauth_user->set_disabledtext('DISABLED');
 $oauth_user->update();
 $t->get_ok('/api/user/profile' =>
-    {Authorization => 'Bearer ' . $access_data->{access_token}})
-  ->status_is(401);
-
-# Should get an error if we try to re-use the same auth code again
-$t->post_ok(
-  '/oauth/access_token' => {Referer => $referer} => form => {
-    client_id     => $oauth_client->{client_id},
-    client_secret => $oauth_client->{secret},
-    code          => $auth_code,
-    grant_type    => 'authorization_code',
-    redirect_uri  => '/oauth/redirect',
-  }
-)->status_is(400)->json_is('/error' => 'invalid_grant');
+    {Authorization => 'Bearer ' . $access_data->{access_token}})->status_is(401);
```

Comment:
I'd like to see a test where you pass in garbage as an access-token to ensure that fails correctly.